### PR TITLE
feat: css var

### PR DESCRIPTION
**Pull request by MadCcc**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
- https://github.com/ant-design/ant-design/discussions/44654
- https://github.com/ant-design/ant-design/discussions/45385

### 💡 Background and solution
已改造 Button 组件 https://preview-45589-ant-design.surge.sh/components/button-cn

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Theme support css variables.     |
| 🇨🇳 Chinese |     主题支持 CSS 变量      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 49b7ab0</samp>

This pull request enhances the theme system of the ant-design library by adding support for CSS variables, dynamic calculations, and theme switching. It also fixes some code previewer errors and removes unused legacy code. The main files affected are `components/theme/*`, `components/config-provider/*`, and `.dumi/theme/builtins/Previewer/CodePreviewer.tsx`.

### <samp>🤖 Generated by Copilot at 49b7ab0</samp>

*  Import and use `ConfigProvider` component to set CSS variables for code box demo ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-54c90701ed595205ea2cd4286b1cd0ee8b2f01f77a23088e0108feda7af2d65fL5-R5), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-54c90701ed595205ea2cd4286b1cd0ee8b2f01f77a23088e0108feda7af2d65fL369-R377))
*  Remove `wrapSSR` function calls from `button.tsx` as they are not needed for code previewer ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L121-R121), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L257-R257), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L269-R269), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-463cd38ab465e1f3b90463a403d8e8399f926361436eaa9adb8a76a8bf9d4140L299-R299))
*  Import and use `unit` function from `@ant-design/cssinjs` to convert numbers to CSS units in `button/style/index.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L2-R6), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L159-R175), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L223-R239), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L241-R257), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L589-R602))
*  Add `paddingBlock`, `paddingBlockLG`, and `paddingBlockSM` properties to `ComponentToken` interface and `buttonPaddingVertical` property to `ButtonToken` interface in `button/style/index.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693R94-R108), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693R157))
*  Use `buttonPaddingVertical` property to define vertical padding of buttons in `button/style/index.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L638-R652), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693R665))
*  Destructure and merge token properties for buttons in `prepareComponentToken` function in `button/style/index.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L673-R692), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-e8f947f22fc66308c74c500faf1cfeab5e8eed33d2ab4a27718770da99224693L683-R745))
*  Comment out `validateBreakpoints` function in `responsiveObserver.ts` as it is not used and causes errors ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-1cd1627a602528d79cb6ef420545e6efe1b84829291921f70f2b92486c98a6acL31-R60))
*  Add `cssVar` property to `ThemeConfig` interface in `config-provider/context.ts` to define options for CSS variables ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaR50-R52))
*  Change `theme` prop of `ConfigProvider` component to use `cssVariables` option and add `cssThemes` prop to define light and dark themes in `config-provider/demo/theme.tsx` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dR12), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-a19cb548fa41fcf5933a871bf0bfffffe96c649e60dcab61be32b7f97f8a3f8dL38-R76))
*  Use `useId` hook to generate a unique id for CSS variables in `config-provider/hooks/useTheme.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3R6), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3R16-R17))
*  Merge `themeConfig.cssVar` and `parentThemeConfig.cssVar` with a default key based on the id in `config-provider/hooks/useTheme.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3R36-R45))
*  Add `cssVar` property to the returned value of `useTheme` hook in `config-provider/hooks/useTheme.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-85ff925f8aa07cbb863b96b5bffc960720ad00516700939d2aa40cde582d41f3R56))
*  Use `DesignTokenProvider` component instead of `DesignTokenContext.Provider` component to pass `memoTheme` prop in `config-provider/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L22-R22), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L572-R572))
*  Add `calc` function to `GenerateStyle` interface and import it from `./util/calc` in `theme/internal.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-34415735daeac3a403575c234bd2a124b11f85d555ec72fe8d345b53f471016fR18), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-34415735daeac3a403575c234bd2a124b11f85d555ec72fe8d345b53f471016fR28))
*  Add `calculator` test suite to test `NumCalculator` and `CSSCalculator` classes from `calc` function in `theme/__tests__/util.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-474a355b4973b5cd9ab68c695f77e8fc28504f72f7af6297c8396151bbfff873R2), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-474a355b4973b5cd9ab68c695f77e8fc28504f72f7af6297c8396151bbfff873R10-R22))
*  Add `DesignTokenProvider` component and its props type to wrap `DesignTokenContext.Provider` component and pass `theme` prop in `theme/context.tsx` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-664e006143798b15378076159fe6e1d295232d6c270cb5d0897ba5d140fe52a4L1-R2), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-664e006143798b15378076159fe6e1d295232d6c270cb5d0897ba5d140fe52a4L32-R53))
*  Remove `LegacyColorPalettes` type and its usage from `theme/interface/index.ts`, `theme/interface/maps/index.ts`, `theme/interface/presetColors.ts`, `theme/themes/dark/index.ts`, and `theme/themes/default/index.ts` as it is deprecated and no longer used ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-7c4eca125442f4e9f24a1897a9eca97ed13cce775c5cb9e7e29f60ee574ede7cL31), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-5291b08922dc58d4f3dce4d94e6fee0d82ce1bd1afe5518fb8cef97c96e0c59fL1-R1), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-5291b08922dc58d4f3dce4d94e6fee0d82ce1bd1afe5518fb8cef97c96e0c59fL39), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-7aa90d5f2df53b9bbe086a8649e868ea0a7de955e36528ecffaf8b32d004e3c0L23-L29), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-202cda9dd4c6ecf0d9435e00f82a40221bc75481f8dcacb4cf563ab70dc0dbe0L3-R3), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-202cda9dd4c6ecf0d9435e00f82a40221bc75481f8dcacb4cf563ab70dc0dbe0L21), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-202cda9dd4c6ecf0d9435e00f82a40221bc75481f8dcacb4cf563ab70dc0dbe0L32-R25), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-f61c122f7c7e832a15a7e4d313ae5b78150e5830c277c34710e55ef759837eb8L4-R4), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-f61c122f7c7e832a15a7e4d313ae5b78150e5830c277c34710e55ef759837eb8L23-R19), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-f61c122f7c7e832a15a7e4d313ae5b78150e5830c277c34710e55ef759837eb8L34-R27))
*  Add `unitless` and `ignore` objects to define which token properties should not have units or should be ignored when generating CSS variables in `theme/useToken.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cR12-R46))
*  Add `realToken` and `cssVarKey` variables to the return type and value of `useToken` and `useCacheToken` hooks to provide the original token and the CSS variable key in `theme/useToken.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cL60-R104), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cL67-R110), [link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-c71eb91865998661d8139ee89317ebd253b704a1fae92b80254b164c05c67e8cL77-R129))
*  Add `AbstractCalculator` class to define the abstract methods for a calculator class in `theme/util/calc/calculator.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-49a0dc142139e67cea36c4adf85e2bf5dd996db079b8a3d26b34c1ea7525f41bR1-R15))
*  Add `CSSCalculator` class to implement the abstract calculator class for CSS values in `theme/util/calc/CSSCalculator.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-d67ed27dfe5fa898d004941bb381fc84b043cbc166978e03b48c21033bd2a86eR1-R57))
*  Add `NumCalculator` class to implement the abstract calculator class for numbers in `theme/util/calc/NumCalculator.ts` ([link](https://github.com/ant-design/ant-design/pull/45589/files?diff=unified&w=0#diff-8375517a8fc9bb5b24cfbfd3722a9c1d0eef49a4a36a8e455d59c1395fc77bc5R1-R33))

### CODE REVIEW

File: .dumi/theme/slots/Content/index.tsx [hunk 1]
```diff
@@ -275,7 +275,9 @@ const Content: React.FC<{ children: ReactNode }> = ({ children }) => {
             </Typography.Paragraph>
           ) : null}
           {!meta.frontmatter.__autoDescription && meta.frontmatter.description}
-          <div style={{ minHeight: 'calc(100vh - 64px)' }}>{children}</div>
+          <ConfigProvider theme={{ cssVar: {} }}>
```
Comment (MadCcc) [inline .dumi/theme/slots/Content/index.tsx:278]: 暂时套在这里，所有组件完成后可以套在 GlobalLayout 里（最外层）

File: components/button/style/index.ts [hunk 2]
```diff
@@ -156,7 +172,7 @@ const genSharedButtonStyle: GenerateStyle<ButtonToken, CSSObject> = (token): CSS
       textAlign: 'center',
       backgroundImage: 'none',
       backgroundColor: 'transparent',
-      border: `${token.lineWidth}px ${token.lineType} transparent`,
+      border: `${unit(token.lineWidth)} ${token.lineType} transparent`,
```
[the remark this answers is not in the document]
Comment (MadCcc) [reply components/button/style/index.ts:175]: calc 会根据上下文转换成纯数字运算或者 css 的 calc；
unit 就是个单纯的加 px 的函数，只判断了字符串。
都是用来兼容 css 变量和纯值的

[the remark this answers is not in the document]
Comment (crazyair) [reply components/button/style/index.ts:175]: 我问的不是功能，是放的位置的区别

File: components/_util/responsiveObserver.ts [hunk 3]
```diff
@@ -62,7 +62,7 @@ const validateBreakpoints = (token: GlobalToken) => {
 };
 
 export default function useResponsiveObserver() {
-  const [, token] = useToken();
+  const [, , , token] = useToken();
```
Comment (kiner-tang) [inline components/_util/responsiveObserver.ts:65]: 感觉参数这么多，是不是换成对象比较好呀

Comment (MadCcc) [reply components/_util/responsiveObserver.ts:65]: 这个编译后是直接取下标的，内部方法还好

Comment (kiner-tang) [reply components/_util/responsiveObserver.ts:65]: 就是用的时候比较费劲，也比较难阅读，或者是不是可以加一个 hooks 包一层

Comment (MadCcc) [reply components/_util/responsiveObserver.ts:65]: 用数组是为了 uglify 的时候可以转化掉，对象不行

File: components/button/style/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MadCcc) [reply components/button/style/index.ts:175]: unit 是 cssinjs 库的标准能力，那边写单测也用得到
calc 也可以放到 cssinjs 库里，不过现在只是在 antd 侧做了判断，是为了处理 antd 的计算逻辑产生的，不一定都用得到

**Review by li-jia-nan — APPROVED** [review]

**Review by kiner-tang — APPROVED** [review]

File: components/button/button.tsx [hunk 4]
```diff
@@ -118,7 +119,8 @@ const InternalButton: React.ForwardRefRenderFunction<
   const { getPrefixCls, autoInsertSpaceInButton, direction, button } = useContext(ConfigContext);
   const prefixCls = getPrefixCls('btn', customizePrefixCls);
 
-  const [wrapSSR, hashId] = useStyle(prefixCls);
+  const [, hashId] = useStyle(prefixCls);
+  const wrapCSSVar = useCSSVar(prefixCls);
```
Comment (crazyair) [inline components/button/button.tsx:123]: 以后组件要引入 useStyle useCSSVar 2个钩子吗，hashId 可以加到 useCSSVar 吗

File: components/button/style/cssVar.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+import { genCSSVarRegister } from '../../theme/internal';
+import { prepareComponentToken } from '.';
+
+export default genCSSVarRegister('Button', prepareComponentToken);
```
Comment (crazyair) [inline components/button/style/cssVar.ts:4]: 这个放到 style/index.ts 里是否可以，就一行代码

File: components/button/style/index.ts [hunk 6]
```diff
@@ -586,7 +599,7 @@ const genSizeButtonStyle = (token: ButtonToken, sizePrefixCls: string = ''): CSS
       [`${componentCls}${sizePrefixCls}`]: {
         fontSize,
         height: controlHeight,
-        padding: `${paddingVertical}px ${buttonPaddingHorizontal}px`,
+        padding: `${unit(buttonPaddingVertical!)} ${unit(buttonPaddingHorizontal!)}`,
```
Comment (crazyair) [inline components/button/style/index.ts:602]: ! 不推荐用吧，可以改下 unit 方法参数非必填

File: components/config-provider/context.ts [hunk 7]
```diff
@@ -47,6 +47,16 @@ export interface ThemeConfig {
...
   inherit?: boolean;
+  cssVar?: {
+    /**
+     * Prefix for css variable, default to `antd`.
+     */
+    prefix?: string;
+    /**
+     * Unique key for theme, should be set manually < react@18.
+     */
+    key?: string;
```
Comment (crazyair) [inline components/config-provider/context.ts:58]: 有没有判断 react 小于18，增加 warning

File: components/theme/context.ts [hunk 8]
```diff
@@ -29,6 +29,10 @@ export interface DesignTokenProviderProps {
   /** Just merge `token` & `override` at top to save perf */
   override: { override: Partial<AliasToken> } & ComponentsToken;
   hashed?: string | boolean;
+  cssVar?: {
+    prefix?: string;
```
Comment (crazyair) [inline components/theme/context.ts:33]: 引用 components/config-provider/context.ts

File: package.json [hunk 9]
```diff
@@ -113,7 +113,7 @@
   ],
   "dependencies": {
     "@ant-design/colors": "^7.0.0",
-    "@ant-design/cssinjs": "^1.17.2",
+    "@ant-design/cssinjs": "^2.0.0-alpha.1",
```
Comment (crazyair) [inline package.json:116]: 这个开会不是说发个 1.2 版本吗

File: components/button/style/index.ts (hunk 6 above)
Comment (MadCcc) [reply components/button/style/index.ts:602]: 这里其实是非空的，类型上有点问题，忘记改了

Comment (MadCcc) [reply components/button/style/index.ts:602]: 我给 unit 兜个底

File: components/button/style/cssVar.ts (hunk 5 above)
Comment (MadCcc) [reply components/button/style/cssVar.ts:4]: 我在考虑和 genComponentStyleHook 合并了，这样也不用提公共方法了

File: components/button/style/index.ts [hunk 10]
```diff
@@ -90,6 +90,21 @@ export interface ComponentToken {
...
+  /**
+   * @desc 大号按钮横向内间距
+   * @descEN Horizontal padding of large button
+   */
+  paddingBlockLG: CSSProperties['paddingInline'];
+  /**
+   * @desc 小号按钮横向内间距
+   * @descEN Horizontal padding of small button
+   */
+  paddingBlockSM: CSSProperties['paddingInline'];
```
Comment (borysyuk) [inline components/button/style/index.ts:107]: @MadCcc @crazyair Seams like we have wrong English descriptions for these tokens.  All `Horizontal` should be replaced by `Vertical`. I have tried to use these tokens and they change vertical padding of `block` button.

**Pull request merged**